### PR TITLE
Register deferred commands immediately before finding command to run

### LINKED ISSUE #4122: WP-CLI 1.2 does not recognize custom commands whose names contain spaces
We have some custom commands whose names contain spaces. That has always worked fine until the 1.2 release of WP-CLI, after which they are no longer recognized.

To reproduce, place the following file in an mu-plugin: https://gist.github.com/philipnewcomer/73b2405aae3ffed4ee3fe2e51fada4b7

Before WP-CLI 1.2:
```
$ wp foo bar
Success: Hello world!
```

After updating to 1.2:
```
$ wp foo bar
Error: 'foo' is not a registered wp command. See 'wp help'.
```

If I simply change the name of the command to `foobar` instead of `foo bar` and run `wp foobar`, it works as expected on 1.2.

### PR DESCRIPTION
**Pull request by danielbachhuber**
Fixes #4122

### CODE REVIEW

**Comment by danielbachhuber** [comment]
@schlessera I don't love this implementation of having `RegisterDeferredCommands::process()` register a hook, and having to introduce a new hook within `WP_CLI\Runner->find_command_to_run()`:

1. It doesn't feel quite right.
2. Every new hook we introduce is some execution point we have to maintain permanently.

Is there a better way of doing this?

**Comment by schlessera** [comment]
Two alternatives I can think of, but not sure they're necessarily better:

1. Provide external access to the code that checks for deferred commands, and directly run it in the correct spot.

2. Add priorities to hooks and run the check for deferred commands at `after_wp_load:9999`.

**Comment by danielbachhuber** [comment]
@schlessera Ok. Let's go for this now, I suppose.

File: php/WP_CLI/Runner.php [hunk 1]
```diff
@@ -247,6 +247,8 @@ private function cmd_starts_with( $prefix ) {
 	public function find_command_to_run( $args ) {
 		$command = \WP_CLI::get_root_command();
 
+		WP_CLI::do_hook( 'find_command_to_run_pre' );
```
Comment (schlessera) [inline php/WP_CLI/Runner.php]: Please add a reference to this hook in the doc block of `WP_CLI::add_hook()`: https://github.com/wp-cli/wp-cli/blob/3ccde5b05bb42c4ce4213e636e96d36e96f24bc3/php/class-wp-cli.php#L206-L215

Comment (danielbachhuber) [inline php/WP_CLI/Runner.php]: 8deb8aab

**Pull request merged**